contracts/modules/royalty/policies/IpPool.sol

```diff
@@ -0,0 +1,225 @@
+        unclaimedRoyaltyTokens = unclaimedTokens;
+
+        _mint(address(this), unclaimedTokens);
+        _mint(ipId, supply - unclaimedTokens);
+    }
+
+    /// @notice Adds a new revenue token to the pool
+    /// @param token The address of the revenue token
+    /// @dev Only callable by the royalty policy LAP
+    function updateIpPoolTokens(address token) external {
+        if (msg.sender != address(ROYALTY_POLICY_LAP)) revert Errors.IpPool__NotRoyaltyPolicyLAP();
+        _tokens.add(token);
+    }
+
+    /// @notice Snapshots the claimable revenue and royalty token amounts
+    /// @return snapshotId The snapshot id
+    function snapshot() external returns (uint256) {
+        if (block.timestamp - lastSnapshotTimestamp < ROYALTY_POLICY_LAP.getSnapshotInterval())
+            revert Errors.IpPool__SnapshotIntervalTooShort();
+
+        uint256 snapshotId = _snapshot();
+        lastSnapshotTimestamp = block.timestamp;
+
+        uint32 unclaimedTokens = unclaimedRoyaltyTokens;
+        unclaimedAtSnapshot[snapshotId] = unclaimedTokens;
+
+        address[] memory tokens = _tokens.values();
+
+        for (uint256 i = 0; i < tokens.length; i++) {
+            uint256 tokenBalance = IERC20(tokens[i]).balanceOf(address(this));
+            if (tokenBalance == 0) {
+                _tokens.remove(tokens[i]);
+                continue;
+            }
+
+            uint256 newRevenue = tokenBalance - claimPoolAmount[tokens[i]] - ancestorsPoolAmount[tokens[i]];
+            if (newRevenue == 0) continue;
+
+            uint256 ancestorsTokens = (newRevenue * unclaimedTokens) / totalSupply();
+            ancestorsPoolAmount[tokens[i]] += ancestorsTokens;
+
+            uint256 claimableTokens = newRevenue - ancestorsTokens;
+            claimableAtSnapshot[snapshotId][tokens[i]] = claimableTokens;
+            claimPoolAmount[tokens[i]] += claimableTokens;
+        }
+
+        emit SnapshotCompleted(snapshotId, block.timestamp, unclaimedTokens);
+
+        return snapshotId;
+    }
+
+    /// @notice Calculates the amount of revenue token claimable by a token holder at certain snapshot
+    /// @param account The address of the token holder
+    /// @param snapshotId The snapshot id
+    /// @param token The revenue token to claim
+    /// @return The amount of revenue token claimable
+    function claimableRevenue(address account, uint256 snapshotId, address token) external view returns (uint256) {
+        return _claimableRevenue(account, snapshotId, token);
+    }
+
+    /// @notice Allows token holders to claim revenue token based on the token balance at certain snapshot
+    /// @param snapshotId The snapshot id
+    /// @param tokens The list of revenue tokens to claim
+    function claimRevenueByTokenBatch(uint256 snapshotId, address[] calldata tokens) external nonReentrant {
+        for (uint256 i = 0; i < tokens.length; i++) {
+            uint256 claimableToken = _claimableRevenue(msg.sender, snapshotId, tokens[i]);
```

Comment:
Why do the loop'ed claim functions use `msg.sender` but claimableRevenue uses `account`? I see that the latter (using `account`) allows calling on behalf of the token owner.

Do we plan to support RT delegation?

Reply:
It's a good point - we could consider permissionless claiming or another type of delegation